File: src/mvmctl/cli/cache.py
"""Cache management commands."""

from __future__ import annotations

from typing import TYPE_CHECKING

import typer
from rich.console import Console

from mvmctl.api import CacheOperation as _CacheOperation

if TYPE_CHECKING:
    from mvmctl.api.cache_operations import CacheOperation
else:
    CacheOperation = _CacheOperation

from mvmctl.cli._completion import _complete_cache_resources
from mvmctl.models.result import ProgressEvent
from mvmctl.utils.cli import handle_errors, mvm_cli

cache_app = typer.Typer(
    help="Cache management",
    no_args_is_help=True,
    add_completion=False,
)


@cache_app.command(name="help", hidden=True)
def help_cmd(ctx: typer.Context) -> None:
    """Show help for the cache command group."""
    typer.echo(ctx.parent.get_help() if ctx.parent else "")
    raise typer.Exit()


@cache_app.command(name="init")
@handle_errors
def cache_init() -> None:
    """Initialize all cache resources."""
    console = Console()
    with console.status("", spinner="dots") as status:

        def _on_progress(event: ProgressEvent) -> None:
            if event.message:
                status.update(f"[dim]{event.message}[/dim]")

        operation_result = CacheOperation.init_all(on_progress=_on_progress)
    if operation_result.is_error:
        mvm_cli.error(operation_result.message)
        raise typer.Exit(code=1)
    mvm_cli.success(operation_result.message)
    item = operation_result.item or {}
    for resource, path in item.items():
        if path:
            mvm_cli.info(f"  {resource}: {path}")


@cache_app.command(name="prune")
@handle_errors
def cache_prune(
    resource: str | None = typer.Argument(
        None,
        help=(
            "Resource to prune: vm, network, image, kernel, binary, misc. "
            "Omit to prune all types."
        ),
        autocompletion=_complete_cache_resources,
    ),
    all_resources: bool = typer.Option(
        False,
        "--all",
        "-a",
        help="Remove ALL items including running VMs, default network, protected assets.",
    ),
    dry_run: bool = typer.Option(
        False,
        "--dry-run",
        help="Show what would be removed without actually removing",
    ),
    force: bool = typer.Option(
        False, "--force", "-f", help="Skip confirmation prompts"
    ),
) -> None:
    """
    Prune cache resources.

    Default behavior prunes all items EXCEPT:
    - RUNNING or STARTING VMs
    - Default network and networks referenced by VMs
    - Default image and images used by VMs
    - Default kernel and kernels used by VMs
    - Default Firecracker binary version

    Use ``--all`` to remove everything including protected items.

    Examples:
        mvm cache prune vm                     # Prune non-running VMs
        mvm cache prune vm --all               # Remove ALL VMs including running
        mvm cache prune image --all            # Remove ALL images including protected
        mvm cache prune network                # Prune unused networks only
        mvm cache prune misc                   # Remove appliance + warm images
        mvm cache prune --all                  # Prune ALL resources including protected
        mvm cache prune --all --force          # Prune all without confirmation

    """
    if resource == "vm":
        if not force and not dry_run:
            mvm_cli.warning("This will remove cached data for all VMs")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        result = CacheOperation.prune_vms(
            dry_run=dry_run, include_all=all_resources
        )
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        removed = result.item or []
        if removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(removed)} VM(s): {', '.join(removed)}"
                )
            else:
                mvm_cli.success(f"Pruned: {', '.join(removed)}")
        else:
            mvm_cli.info("No VMs to prune")

    elif resource == "network":
        if not force and not dry_run:
            mvm_cli.warning("This will remove cached data for all networks")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        result = CacheOperation.prune_networks(
            dry_run=dry_run, include_all=all_resources
        )
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        removed = result.item or []
        if removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(removed)} network(s): {', '.join(removed)}"
                )
            else:
                mvm_cli.success(f"Pruned: {', '.join(removed)}")
        else:
            mvm_cli.info("No networks to prune")

    elif resource == "image":
        if not force and not dry_run:
            mvm_cli.warning("This will remove cached data for all images")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        result = CacheOperation.prune_images(
            dry_run=dry_run, include_all=all_resources
        )
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        removed = result.item or []
        if removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(removed)} image(s): {', '.join(removed)}"
                )
            else:
                mvm_cli.success(f"Pruned: {', '.join(removed)}")
        else:
            mvm_cli.info("No images to prune")

    elif resource == "kernel":
        if not force and not dry_run:
            mvm_cli.warning("This will remove cached data for all kernels")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        result = CacheOperation.prune_kernels(
            dry_run=dry_run, include_all=all_resources
        )
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        removed = result.item or []
        if removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(removed)} kernel(s): {', '.join(removed)}"
                )
            else:
                mvm_cli.success(f"Pruned: {', '.join(removed)}")
        else:
            mvm_cli.info("No kernels to prune")

    elif resource == "binary":
        if not force and not dry_run:
            mvm_cli.warning("This will remove cached data for all binaries")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        result = CacheOperation.prune_binaries(
            dry_run=dry_run, include_all=all_resources
        )
        if result.is_error:
            mvm_cli.error(result.message)
            raise typer.Exit(code=1)
        removed = result.item or []
        if removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(removed)} binaries: {', '.join(removed)}"
                )
            else:
                mvm_cli.success(f"Pruned: {', '.join(removed)}")
        else:
            mvm_cli.info("No binaries to prune")

    elif resource == "misc":
        if not force and not dry_run:
            mvm_cli.warning(
                "This will remove cached data (appliance folder, warm images)"
            )
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                raise typer.Exit()
        misc_op_result = CacheOperation.prune_misc(dry_run=dry_run)
        if misc_op_result.is_error:
            mvm_cli.error(misc_op_result.message)
            raise typer.Exit(code=1)
        misc_result = misc_op_result.item or {}
        if misc_result.get("appliance"):
            if dry_run:
                mvm_cli.info("[DRY RUN] Would remove appliance folder")
            else:
                mvm_cli.success("Removed: appliance folder")
        if misc_result.get("warm_images"):
            if dry_run:
                mvm_cli.info("[DRY RUN] Would remove warm images (ready pool)")
            else:
                mvm_cli.success("Removed: warm images (ready pool)")
        if not misc_result.get("appliance") and not misc_result.get(
            "warm_images"
        ):
            mvm_cli.info("No misc cache to prune")

    elif resource is None or all_resources:
        if not all_resources:
            mvm_cli.error(
                "No resource specified. Use --all to prune all resource types."
            )
            mvm_cli.info(
                "Valid resources: vm, network, image, kernel, binary, misc"
            )
            mvm_cli.info("Or use: mvm cache prune --all  # Prune all types")
            raise typer.Exit(code=1)

        if dry_run:
            mvm_cli.info("[DRY RUN] The following would be removed:")
            mvm_cli.info("  - ALL VMs (including RUNNING and STARTING)")
            mvm_cli.info("  - ALL networks (including default)")
            mvm_cli.info("  - ALL images (including default)")
            mvm_cli.info("  - ALL kernels (including default)")
            mvm_cli.info("  - ALL binaries (including default)")
            mvm_cli.info("  - Appliance folder (libguestfs cache)")
            mvm_cli.info("  - Warm images (tmpfs ready pool)")
        elif not force and not dry_run:
            mvm_cli.warning(
                "This will remove ALL cache resources INCLUDING protected items:"
            )
            mvm_cli.info("  - ALL VMs (including RUNNING and STARTING)")
            mvm_cli.info("  - ALL networks (including default)")
            mvm_cli.info("  - ALL images (including default)")
            mvm_cli.info("  - ALL kernels (including default)")
            mvm_cli.info("  - ALL binaries (including default)")
            mvm_cli.info("  - Appliance folder (libguestfs cache)")
            mvm_cli.info("  - Warm images (tmpfs ready pool)")
            mvm_cli.info("")
            if not typer.confirm("Continue?", default=True):
                mvm_cli.info("Aborted")
                raise typer.Exit()

        prune_op_result = CacheOperation.prune_all(
            dry_run=dry_run, include_all=True
        )
        if prune_op_result.is_error:
            mvm_cli.error(prune_op_result.message)
            raise typer.Exit(code=1)

        prune_item = prune_op_result.item
        if prune_item and prune_item.pruned_ids:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(prune_item.pruned_ids)} item(s)"
                )
            else:
                mvm_cli.success("Pruned")

        if prune_item and prune_item.failed_ids:
            mvm_cli.warning(
                f"Failed to prune {len(prune_item.failed_ids)} item(s): "
                f"{', '.join(prune_item.failed_ids)}"
            )

        if prune_item and prune_item.had_running_vms:
            mvm_cli.info(
                "Note: running or starting VMs were present during prune"
            )

    else:
        mvm_cli.error(f"Unknown resource: {resource}")
        mvm_cli.info(
            "Valid resources: vm, network, image, kernel, binary, misc"
        )
        mvm_cli.info(
            "Or use: mvm cache prune --all  # Prune all types including protected"
        )
        raise typer.Exit(code=1)


@cache_app.command(name="clean")
@handle_errors
def cache_clean(
    dry_run: bool = typer.Option(
        False,
        "--dry-run",
        help="Show what would be removed without actually removing",
    ),
    force: bool = typer.Option(
        False, "--force", "-f", help="Skip confirmation prompts"
    ),
) -> None:
    """Completely clean all cache — prune everything, host clean, remove cache dir.

    This is the "nuclear option" for cache cleanup. It:
    1. Prunes ALL resources (VMs, networks, images, kernels, binaries, misc)
    2. Cleans host networking (TAPs, bridges, iptables chains)
    3. Removes the entire cache directory at ~/.cache/mvmctl

    Examples:
        mvm cache clean                # Clean all cache (with confirmation)
        mvm cache clean --dry-run      # Preview what would be removed
        mvm cache clean --force        # Clean without confirmation
    """
    if dry_run:
        mvm_cli.info("[DRY RUN] The following would be removed:")
        mvm_cli.info("  - ALL VMs (including RUNNING and STARTING)")
        mvm_cli.info("  - ALL networks (including default)")
        mvm_cli.info("  - ALL images (including default)")
        mvm_cli.info("  - ALL kernels (including default)")
        mvm_cli.info("  - ALL binaries (including default)")
        mvm_cli.info("  - Appliance folder (libguestfs cache)")
        mvm_cli.info("  - Warm images (tmpfs ready pool)")
        mvm_cli.info("  - Host networking (TAPs, bridges, iptables chains)")
        mvm_cli.info("  - Entire cache directory (~/.cache/mvmctl)")
    elif not force and not dry_run:
        mvm_cli.warning("This will COMPLETELY remove ALL cache data INCLUDING:")
        mvm_cli.info("  - ALL VMs (including RUNNING and STARTING)")
        mvm_cli.info("  - ALL networks (including default)")
        mvm_cli.info("  - ALL images (including default)")
        mvm_cli.info("  - ALL kernels (including default)")
        mvm_cli.info("  - ALL binaries (including default)")
        mvm_cli.info("  - Appliance folder (libguestfs cache)")
        mvm_cli.info("  - Warm images (tmpfs ready pool)")
        mvm_cli.info("  - Host networking (TAPs, bridges, iptables chains)")
        mvm_cli.info("  - Entire cache directory (~/.cache/mvmctl)")
        mvm_cli.info("")
        if not typer.confirm("Continue?", default=True):
            raise typer.Exit()

    op_result = CacheOperation.clean(dry_run=dry_run)
    if op_result.is_error:
        mvm_cli.error(op_result.message)
        raise typer.Exit(code=1)

    result = op_result.item
    if result:
        prune = result.prune_result
        if prune.pruned_ids:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would prune {len(prune.pruned_ids)} item(s)"
                )
            else:
                mvm_cli.success("Pruned")

        if prune.failed_ids:
            mvm_cli.warning(
                f"Failed to prune {len(prune.failed_ids)} item(s): "
                f"{', '.join(prune.failed_ids)}"
            )

        if prune.had_running_vms:
            mvm_cli.info(
                "Note: running or starting VMs were present during clean"
            )

        if result.cache_dir_removed:
            if dry_run:
                mvm_cli.info(
                    f"[DRY RUN] Would remove cache directory: {result.cache_dir}"
                )
            else:
                mvm_cli.success(f"Removed: {result.cache_dir}")
        else:
            mvm_cli.info("Cache directory was already empty")


__all__ = ["cache_app"]

File: internal/cli/cache.go
// Package cli — cache management commands
package cli

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"

	"mvmctl/internal/cli/common"
	"mvmctl/internal/infra"
	"mvmctl/internal/infra/event"
	"mvmctl/internal/lib/system"
	"mvmctl/pkg/api"

	"github.com/spf13/cobra"
)

// NewCacheCmd creates the cache command and its subcommands.
func NewCacheCmd(cacheAPI api.CacheAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "cache",
		Short: "Cache management",
	}

	cmd.AddCommand(newCacheInitCmd(cacheAPI))
	cmd.AddCommand(newCachePruneCmd(cacheAPI))
	cmd.AddCommand(newCacheCleanCmd(cacheAPI))
	return cmd
}

func newCacheInitCmd(cacheAPI api.CacheAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "init",
		Short: "Initialize all cache resources",
		RunE: func(cmd *cobra.Command, args []string) error {
		// Uses progress callback for status updates during cache init.
			prog := common.NewProgress()
			prog.Start("Initializing cache...")

			item, err := cacheAPI.CacheInitAll(cmd.Context(), func(e event.Progress) {
				if e.Message != "" {
					prog.UpdateText(e.Message)
				}
			})

			prog.Stop()

			if err != nil {
				return err
			}

			common.Cli.Success("Cache initialized successfully")
			for _, dir := range item.Directories {
				common.Cli.Info(fmt.Sprintf("  %s", dir))
			}
			if item.GuestfsAppliance != "" {
				common.Cli.Info(fmt.Sprintf("  guestfs: %s", item.GuestfsAppliance))
			}
			if item.GuestfsKernel != "" {
				common.Cli.Info(fmt.Sprintf("  kernel: %s", item.GuestfsKernel))
			}

			return nil
		},
	}

	return cmd
}

// resourceDisplayName returns the display name for a resource type in messages.
func resourceDisplayName(resource string) string {
	switch resource {
	case "vm":
		return "VM"
	case "network":
		return "network"
	case "image":
		return "image"
	case "kernel":
		return "kernel"
	case "binary":
		return "binary"
	default:
		return resource
	}
}

// resourceDisplayNamePlural returns the plural display name for a resource type.
func resourceDisplayNamePlural(resource string) string {
	switch resource {
	case "vm":
		return "VMs"
	case "network":
		return "networks"
	case "image":
		return "images"
	case "kernel":
		return "kernels"
	case "binary":
		return "binaries"
	default:
		return resource + "s"
	}
}

// pruneResource handles pruning a specific resource type.
func pruneResource(
	cacheAPI api.CacheAPI,
	cmd *cobra.Command,
	resource string,
	dryRun bool,
	allResources bool,
	force bool,
) error {
	if !force && !dryRun {
		common.Cli.Warning(fmt.Sprintf("This will remove cached data for all %s", resourceDisplayNamePlural(resource)))
		common.Cli.Info("")
		confirmed, pErr := common.Cli.PromptConfirm(cmd.Context(), "Continue?", true)
		if pErr != nil {
			return pErr
		}
		if !confirmed {
			return nil
		}
	}

	var removed []string

	switch resource {
	case "image":
		ids, err := cacheAPI.CachePruneImages(cmd.Context(), dryRun, allResources)
		if err != nil {
			return err
		}
		removed = ids
	case "binary":
		ids, err := cacheAPI.CachePruneBinaries(cmd.Context(), dryRun, allResources)
		if err != nil {
			return err
		}
		removed = ids
	default:
		switch resource {
		case "vm":
			opResult := cacheAPI.CachePruneVMs(cmd.Context(), dryRun, allResources)
			if opResult == nil {
				return nil
			}
			if opResult.IsError() {
				return opResult.ToError()
			}
			if item, ok := opResult.Item.([]string); ok {
				removed = item
			}
		case "network":
			ids, err := cacheAPI.CachePruneNetworks(cmd.Context(), dryRun, allResources)
			if err != nil {
				return err
			}
			removed = ids
		case "kernel":
			ids, err := cacheAPI.CachePruneKernels(cmd.Context(), dryRun, allResources)
			if err != nil {
				return err
			}
			removed = ids
		}
	}

	if len(removed) > 0 {
		if dryRun {
			if resource == "binary" {
				common.Cli.Info(
					fmt.Sprintf("[DRY RUN] Would prune %d binaries: %s", len(removed), strings.Join(removed, ", ")),
				)
			} else {
				displayName := resourceDisplayName(resource)
				common.Cli.Info(
					fmt.Sprintf(
						"[DRY RUN] Would prune %d %s(s): %s",
						len(removed),
						displayName,
						strings.Join(removed, ", "),
					),
				)
			}
		} else {
			common.Cli.Success(fmt.Sprintf("Pruned: %s", strings.Join(removed, ", ")))
		}
	} else {
		plural := resourceDisplayNamePlural(resource)
		common.Cli.Info(fmt.Sprintf("No %s to prune", plural))
	}

	return nil
}

func pruneMisc(cacheAPI api.CacheAPI, cmd *cobra.Command, dryRun bool, force bool) error {
	if !force && !dryRun {
		common.Cli.Warning("This will remove cached data (appliance folder, warm images)")
		common.Cli.Info("")
		confirmed, pErr := common.Cli.PromptConfirm(cmd.Context(), "Continue?", true)
		if pErr != nil {
			return pErr
		}
		if !confirmed {
			return nil
		}
	}

	miscMap, err := cacheAPI.CachePruneMisc(cmd.Context(), dryRun)
	if err != nil {
		return err
	}

	if miscMap == nil {
		common.Cli.Info("No misc cache to prune")
		return nil
	}

	applianceRemoved, _ := miscMap["appliance"].(bool)
	warmRemoved, _ := miscMap["warm_images"].(bool)

	if applianceRemoved {
		if dryRun {
			common.Cli.Info("[DRY RUN] Would remove appliance folder")
		} else {
			common.Cli.Success("Removed: appliance folder")
		}
	}

	if warmRemoved {
		if dryRun {
			common.Cli.Info("[DRY RUN] Would remove warm images (ready pool)")
		} else {
			common.Cli.Success("Removed: warm images (ready pool)")
		}
	}

	if !applianceRemoved && !warmRemoved {
		common.Cli.Info("No misc cache to prune")
	}

	return nil
}

func pruneAll(cacheAPI api.CacheAPI, cmd *cobra.Command, dryRun bool, force bool) error {
	if dryRun {
		common.Cli.Info("[DRY RUN] The following would be removed:")
		common.Cli.Info("  - ALL VMs (including RUNNING and STARTING)")
		common.Cli.Info("  - ALL networks (including default)")
		common.Cli.Info("  - ALL images (including default)")
		common.Cli.Info("  - ALL kernels (including default)")
		common.Cli.Info("  - ALL binaries (including default)")
		common.Cli.Info("  - Appliance folder (libguestfs cache)")
		common.Cli.Info("  - Warm images (tmpfs ready pool)")
	} else if !force {
		common.Cli.Warning("This will remove ALL cache resources INCLUDING protected items:")
		common.Cli.Info("  - ALL VMs (including RUNNING and STARTING)")
		common.Cli.Info("  - ALL networks (including default)")
		common.Cli.Info("  - ALL images (including default)")
		common.Cli.Info("  - ALL kernels (including default)")
		common.Cli.Info("  - ALL binaries (including default)")
		common.Cli.Info("  - Appliance folder (libguestfs cache)")
		common.Cli.Info("  - Warm images (tmpfs ready pool)")
		common.Cli.Info("")
		confirmed, pErr := common.Cli.PromptConfirm(cmd.Context(), "Continue?", true)
		if pErr != nil {
			return pErr
		}
		if !confirmed {
			common.Cli.Info("Aborted")
			return nil
		}
	}

	pruneItem, err := cacheAPI.CachePruneAll(cmd.Context(), dryRun, true)
	if err != nil {
		return err
	}
	if pruneItem != nil {
		if len(pruneItem.PrunedIDs) > 0 {
			if dryRun {
				common.Cli.Info(fmt.Sprintf("[DRY RUN] Would prune %d item(s)", len(pruneItem.PrunedIDs)))
			} else {
				common.Cli.Success("Pruned")
			}
		}

		if len(pruneItem.FailedIDs) > 0 {
			common.Cli.Warning(
				fmt.Sprintf(
					"Failed to prune %d item(s): %s",
					len(pruneItem.FailedIDs),
					strings.Join(pruneItem.FailedIDs, ", "),
				),
			)
		}

		if pruneItem.HadRunningVMs {
			common.Cli.Info("Note: running or starting VMs were present during prune")
		}
	}

	return nil
}

func newCachePruneCmd(cacheAPI api.CacheAPI) *cobra.Command {
	var allResources bool
	var dryRun bool
	var force bool

	cmd := &cobra.Command{
		Use:               "prune [resource]",
		Short:             "Prune cache resources",
		ValidArgsFunction: completeCacheResources,
		Long: `Prune cache resources.

Default behavior prunes all items EXCEPT:
- RUNNING or STARTING VMs
- Default network and networks referenced by VMs
- Default image and images used by VMs
- Default kernel and kernels used by VMs
- Default Firecracker binary version

Use --all to remove everything including protected items.

Examples:
  mvm cache prune vm                     # Prune non-running VMs
  mvm cache prune vm --all               # Remove ALL VMs including running
  mvm cache prune image --all            # Remove ALL images including protected
  mvm cache prune network                # Prune unused networks only
  mvm cache prune misc                   # Remove appliance + warm images
  mvm cache prune --all                  # Prune ALL resources including protected
  mvm cache prune --all --force          # Prune all without confirmation`,
		Args: cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			var resource string
			if len(args) > 0 {
				resource = args[0]
			}

			switch resource {
			case "vm":
				return pruneResource(cacheAPI, cmd, "vm", dryRun, allResources, force)
			case "network":
				return pruneResource(cacheAPI, cmd, "network", dryRun, allResources, force)
			case "image":
				return pruneResource(cacheAPI, cmd, "image", dryRun, allResources, force)
			case "kernel":
				return pruneResource(cacheAPI, cmd, "kernel", dryRun, allResources, force)
			case "binary":
				return pruneResource(cacheAPI, cmd, "binary", dryRun, allResources, force)
			case "misc":
				return pruneMisc(cacheAPI, cmd, dryRun, force)
			case "":
				if allResources {
					return pruneAll(cacheAPI, cmd, dryRun, force)
				}
				common.Cli.Error("No resource specified. Use --all to prune all resource types.")
				common.Cli.Info("Valid resources: vm, network, image, kernel, binary, misc")
				common.Cli.Info("Or use: mvm cache prune --all  # Prune all types")
				return fmt.Errorf("no resource specified")
			default:
				// Unknown resource with --all should prune all resources.
				if allResources {
					return pruneAll(cacheAPI, cmd, dryRun, force)
				}
				common.Cli.Error(fmt.Sprintf("Unknown resource: %s", resource))
				common.Cli.Info("Valid resources: vm, network, image, kernel, binary, misc")
				common.Cli.Info("Or use: mvm cache prune --all  # Prune all types including protected")
				return fmt.Errorf("unknown resource: %s", resource)
			}
		},
	}

	cmd.Flags().BoolVarP(&allResources, "all", "a", false,
		"Remove ALL items including running VMs, default network, protected assets.")
	cmd.Flags().BoolVar(&dryRun, "dry-run", false,
		"Show what would be removed without actually removing")
	cmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation prompts")
	return cmd
}

// runCacheCleanWithSudo re-invokes "mvm cache clean" with sudo.
func runCacheCleanWithSudo(ctx context.Context) error {
	mvmBin, err := exec.LookPath(infra.CLIName)
	if err != nil {
		mvmBin, err = os.Executable()
		if err != nil {
			mvmBin = infra.CLIName
		}
	}

	common.Cli.Info("")
	common.Cli.Info("Running cache clean with sudo...")
	result, err := system.DefaultRunner.Run(ctx, []string{"sudo", mvmBin, "cache", "clean"}, system.RunCmdOpts{
		Capture: false,
		Check:   false,
	})
	if err != nil || !result.Success() {
		return fmt.Errorf("cache clean with sudo failed (exit %d)", result.ExitCode)
	}
	return nil
}

func newCacheCleanCmd(cacheAPI api.CacheAPI) *cobra.Command {
	var dryRun bool
	var force bool

	cmd := &cobra.Command{
		Use:   "clean",
		Short: "Completely clean all cache — prune everything, host clean, remove cache dir",
		Long: `Completely clean all cache — prune everything, host clean, remove cache dir.

This is the "nuclear option" for cache cleanup. It:
1. Prunes ALL resources (VMs, networks, images, kernels, binaries, misc)
2. Cleans host networking (TAPs, bridges, iptables chains)
3. Removes the entire cache directory at ~/.cache/mvmctl

Examples:
  mvm cache clean                # Clean all cache (with confirmation)
  mvm cache clean --dry-run      # Preview what would be removed
  mvm cache clean --force        # Clean without confirmation`,
		RunE: func(cmd *cobra.Command, args []string) error {
			// Check privileges early — before the destructive confirmation.
			if err := cacheAPI.CacheCheckPrivileges("/usr/sbin/ip", "clean cache"); err != nil {
				if cacheAPI.CacheSessionHasGroup() {
					// Group is active but something else is wrong (missing binary, etc.)
					// Return the original error — let Cobra's root error handler display it.
					return err
				}
				// Session doesn't have the group — offer sudo
				sudoConfirmed, pErr := common.Cli.PromptConfirm(
					cmd.Context(),
					"Elevated privileges required. Run with sudo instead?",
					true,
				)
				if pErr != nil {
					return pErr
				}
				if !sudoConfirmed {
					common.Cli.Info("Aborted")
					return nil
				}
				return runCacheCleanWithSudo(cmd.Context())
			}

			if dryRun {
				common.Cli.Info("[DRY RUN] The following would be removed:")
				common.Cli.Info("  - ALL VMs (including RUNNING and STARTING)")
				common.Cli.Info("  - ALL networks (including default)")
				common.Cli.Info("  - ALL images (including default)")
				common.Cli.Info("  - ALL kernels (including default)")
				common.Cli.Info("  - ALL binaries (including default)")
				common.Cli.Info("  - Appliance folder (libguestfs cache)")
				common.Cli.Info("  - Warm images (tmpfs ready pool)")
				common.Cli.Info("  - Host networking (TAPs, bridges, iptables chains)")
				common.Cli.Info("  - Entire cache directory (~/.cache/mvmctl)")
			} else if !force {
				common.Cli.Warning("This will COMPLETELY remove ALL cache data INCLUDING:")
				common.Cli.Info("  - ALL VMs (including RUNNING and STARTING)")
				common.Cli.Info("  - ALL networks (including default)")
				common.Cli.Info("  - ALL images (including default)")
				common.Cli.Info("  - ALL kernels (including default)")
				common.Cli.Info("  - ALL binaries (including default)")
				common.Cli.Info("  - Appliance folder (libguestfs cache)")
				common.Cli.Info("  - Warm images (tmpfs ready pool)")
				common.Cli.Info("  - Host networking (TAPs, bridges, iptables chains)")
				common.Cli.Info("  - Entire cache directory (~/.cache/mvmctl)")
				common.Cli.Info("")
				cleanConfirmed, pErr := common.Cli.PromptConfirm(cmd.Context(), "Continue?", true)
				if pErr != nil {
					return pErr
				}
				if !cleanConfirmed {
					return nil
				}
			}

			cleanResult, err := cacheAPI.CacheClean(cmd.Context(), dryRun)
			if err != nil {
				return err
			}
			if cleanResult != nil {
				prune := cleanResult.PruneResult

				if len(prune.PrunedIDs) > 0 {
					if dryRun {
						common.Cli.Info(fmt.Sprintf("[DRY RUN] Would prune %d item(s)", len(prune.PrunedIDs)))
					} else {
						common.Cli.Success("Pruned")
					}
				}

				if len(prune.FailedIDs) > 0 {
					common.Cli.Warning(
						fmt.Sprintf(
							"Failed to prune %d item(s): %s",
							len(prune.FailedIDs),
							strings.Join(prune.FailedIDs, ", "),
						),
					)
				}

				if prune.HadRunningVMs {
					common.Cli.Info("Note: running or starting VMs were present during clean")
				}

				if cleanResult.CacheDirRemoved {
					if dryRun {
						common.Cli.Info(fmt.Sprintf("[DRY RUN] Would remove cache directory: %s", cleanResult.CacheDir))
					} else {
						common.Cli.Success(fmt.Sprintf("Removed: %s", cleanResult.CacheDir))
					}
				} else {
					common.Cli.Info("Cache directory was already empty")
				}
			}
			return nil
		},
	}

	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Show what would be removed without actually removing")
	cmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation prompts")
	return cmd
}
